core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -215,10 +219,34 @@ private void acquireConnectionAndExecute0(ClientRequestContext ctx, Endpoint end
         }
     }
 
-    private ProxyConfig getProxyConfig(SessionProtocol protocol, Endpoint endpoint) {
+    private ProxyConfig getProxyConfig(SessionProtocol protocol, Endpoint endpoint, ClientRequestContext ctx) {
```

Comment:
We have to recreate `ProxyConfig` with the new proxy address because the address is resolved:
```java
if (proxyConfig.proxyAddress() != null) {
    final Future<InetSocketAddress> resolveFuture = addressResolverGroup.getResolver(
            ctx.eventLoop().withoutContext()).resolve(proxyConfig.proxyAddress());

    resolveFuture.addListener(future -> {
        if (future.isSuccess()) {
            final ProxyConfig newProxyConfig = proxyConfig.withNewProxyAddress(
                    (InetSocketAddress) future.get());
            ...
        }
    }
}
```

Because getting the `ProxyConfig` is now asynchronous, we have to change the signature of this method.
We have two options:
- Returning `CompletableFuture<ProxyConfig>` instead of `ProxyConfig`
  ```java
  
  // in HttpClientDelegate.execute()
  try {
      final CompletableFuture<ProxyConfig> future = getProxyConfig(protocol, endpoint, ctx);
      future.handle(....)
  } catch (Throwable t) {
      return earlyFailedResponse(t, ctx);
  }
  ```
- Using a callback that is executed after resolution like we do for [resolving address](https://github.com/line/armeria/blob/ab4327765b2cde799818e52d46926809233ca246/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L122-L129)
  ```java
  private void resolveProxyConfig(..., 
      BiConsumer<@Nullable ProxyConig, @Nullable Throwable> onComplete) {
      final ProxyConfig proxyConfig = factory.proxyConfigSelector().select(protocol, endpoint);
        requireNonNull(proxyConfig, "proxyConfig");

        ...
  }
  ```

Please let me know if you need further information. 😉

Reply:
@minwoox , Thanks for your comments 🙇‍♂️ 
I made a new commit with following second option!

Because `ProxyConfig` will be resolved in other thread, so the method `maybeHAProxy()` will get `capturedProxyAddress` instead of `ServiceRequestContext`. 
Also, `res` instance will be returned right away even if future don't completed. 
So I used the method `earlyFailed(...)` instead of `earlyfailedResponse(...)`. 

When you have time, Please take another look 🙇‍♂️